### PR TITLE
Specialize images types to support multiple sample types

### PR DESCRIPTION
**Pull request by alan-baker**
Specialize OpenCL image types in clspv to encode the way they are used to generate correct SPIR-V and support multiple sample types.

A new pass performs type mutation replacing types with specializations like `<base image type>.<sample type>[.sampled]`. To support legacy code generation and tests, if no specializing builtin can be found, the specialized type defaults to a sampled float image.

Adds a new set of utility functions for identifying builtins rather relying on replicating strings in multiple files. Updated some arg utility functions to identify specialized images correctly as images.

Adds basic support for code generation of uint and int images. Currently can handle 2d and 3d sampled reads and 2d and 3d writes. Int images specially generate a signed 32-bit integer and a vec4 of a signed 32-bit integer for use. Bitcasts to/from unsigned 32-bit values are used to appropriately generate the exact type match for image writes or from image reads.

Many new tests.

### CODE REVIEW

**Comment by alan-baker** [comment]
Fixes #286

File: lib/SpecializeImageTypes.cpp [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+  }
+
+  return nullptr;
+}
+
+Type *SpecializeImageTypesPass::RemapUse(Value *value, unsigned operand_no) {
+  if (CallInst *call = dyn_cast<CallInst>(value)) {
+    auto called = call->getCalledFunction();
+    if (IsSampledImageRead(called) || IsImageWrite(called)) {
+      // Specialize the image type based on the it's usage in the builtin.
```
Comment (dneto0) [inline lib/SpecializeImageTypes.cpp]: grammar in comment

File: lib/SpecializeImageTypes.cpp [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+
+  return changed;
+}
+
+Type *SpecializeImageTypesPass::RemapType(Argument *arg) {
+  for (auto &U : arg->uses()) {
+    if (auto new_type = RemapUse(U.getUser(), U.getOperandNo())) {
+      return new_type;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply lib/SpecializeImageTypes.cpp]: Vulkan can't do that with a single binding. An image must either be used with a sampler or not. Theoretically we could generate two bindings for one kernel arg, but that sounds awkward.

File: lib/SPIRVProducerPass.cpp [hunk 3]
```diff
@@ -1820,13 +1840,40 @@ void SPIRVProducerPass::GenerateSPIRVTypes(LLVMContext &Context,
...
-          uint32_t SampledTyID = lookupType(Type::getFloatTy(Context));
+          uint32_t ImageTyID = nextID++;
+          uint32_t SampledTyID = 0;
+          if (STy->getName().contains(".float")) {
+            SampledTyID = lookupType(Type::getFloatTy(Context));
+          } else if (STy->getName().contains(".uint")) {
+            SampledTyID = lookupType(Type::getInt32Ty(Context));
+          } else if (STy->getName().contains(".int")) {
+            // Generate a signed 32-bit integer
+            int32ID = nextID++;
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply lib/SPIRVProducerPass.cpp]: Good catch. Thought I had done that! I'll add the test.

File: lib/ArgKind.cpp [hunk 4]
```diff
@@ -150,6 +150,34 @@ bool IsImageType(llvm::Type *type, llvm::Type **struct_type_ptr) {
   return isImageType;
 }
 
+bool IsFloatImageType(Type *type) {
+  return !IsIntImageType(type) && !IsUintImageType(type);
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply lib/ArgKind.cpp]: I'll add the image check.

**Pull request merged**